src/named_tuple.cr

```diff
@@ -38,6 +38,46 @@ struct NamedTuple
     options
   end
 
+  # Creates a named tuple from the given hash, with elements casted to the given types. See `#from`.
+  #
+  # ```
+  # NamedTuple(foo: String, bar: Int64).from({:foo => "world", :bar => 2})       # => {foo: "world", bar: 2}
+  # NamedTuple(foo: String, bar: Int64).from({"foo" => "world", "bar" => 2})     # => {foo: "world", bar: 2}
+  # NamedTuple(foo: String, bar: Int64).from({:foo => "world", :bar => 2}).class # => {foo: String, bar: Int64}
+  # ```
+  def self.from(hash : Hash)
+    {% begin %}
+    new(**{{@type.type_vars.first}}).from(hash)
+    {% end %}
+  end
+
+  # Expects to be called on a named tuple whose values are types, creates a tuple from the given hash,
+  # with types casted appropriately. The hash keys must be either symbols or strings.
+  #
+  # This allows you to easily pass a hash as individual named arguments to a method.
+  #
+  # ```
+  # def speak_about(thing : String, n : Int64)
+  #   "I see #{n} #{thing}s"
+  # end
+  #
+  # data = JSON.parse(%({"thing": "world", "n": 2})).as_h
+  # speak_about(**{thing: String, n: Int64}.from(data)) # => "I see 2 worlds"
+  # ```
+  def from(hash : Hash)
+    if size != hash.size
+      raise ArgumentError.new("Expected a hash with #{size} keys but one with #{hash.size} keys was given.")
+    end
```

Comment:
I'm a bit unsure about this requirement, I did it mostly for consistency with `Tuple#from` where, it makes more sense. The idea is that people are notified if their assumptions change and/or explicitly discard unwanted keys/items with `Hash#select(*keys)`/`Array#first(n)`